File: scripts/matcher.py
import math
from typing import List, Dict, Any, Tuple, Optional
try:
    from sklearn.feature_extraction.text import TfidfVectorizer
    from sklearn.metrics.pairwise import cosine_similarity
    SKLEARN_AVAILABLE = True
except ImportError:
    SKLEARN_AVAILABLE = False

class Matcher:
    def __init__(self, use_embedding: bool = False, embedding_model=None):
        self.use_embedding = use_embedding and (embedding_model is not None)
        self.embedding_model = embedding_model
        if not SKLEARN_AVAILABLE and not self.use_embedding:
            raise RuntimeError("scikit-learn is required for TF‑IDF matching. Install with: pip install scikit-learn")

    def compute_similarity(self, text1: str, text2: str) -> float:
        """Return cosine similarity between two texts (0-1)."""
        if self.use_embedding:
            emb1 = self.embedding_model.encode([text1])[0]
            emb2 = self.embedding_model.encode([text2])[0]
            return float(cosine_similarity([emb1], [emb2])[0][0])
        else:
            vectorizer = TfidfVectorizer().fit([text1, text2])
            tfidf = vectorizer.transform([text1, text2])
            return float(cosine_similarity(tfidf[0:1], tfidf[1:2])[0][0])

    def analyze_overlap(self, target: Dict[str, Any], local_list: List[Dict[str, Any]]) -> List[Dict]:
        """Return overlap info for each local capability."""
        target_text = f"{target.get('name','')} {target.get('description','')}"
        results = []
        for local in local_list:
            local_text = f"{local.get('name','')} {local.get('description','')}"
            sim = self.compute_similarity(target_text, local_text)
            if sim > 0.85:
                level = "high"
            elif sim > 0.6:
                level = "partial"
            else:
                level = "none"
            results.append({
                "local_name": local["name"],
                "local_type": local["type"],
                "agent": local.get("agent", "unknown"),
                "similarity": round(sim, 3),
                "level": level,
                "reason": self._explain_overlap(target, local, sim)
            })
        # sort by similarity descending
        results.sort(key=lambda x: x["similarity"], reverse=True)
        return results

    def _explain_overlap(self, target: Dict, local: Dict, sim: float) -> str:
        if sim > 0.85:
            return "Likely duplicate functionality"
        elif sim > 0.6:
            return "Partial functional overlap"
        else:
            return "No significant overlap"

    def analyze_synergy(self, target: Dict, local_list: List[Dict]) -> List[Dict]:
        """Heuristic synergy detection based on keyword flow."""
        synergies = []
        target_desc = target.get("description", "").lower()
        for local in local_list:
            local_desc = local.get("description", "").lower()
            # Simple pipeline: target fetches/reads -> local sends/writes
            if any(kw in target_desc for kw in ["fetch", "get", "read", "list", "query"]) and \
               any(kw in local_desc for kw in ["send", "post", "write", "create", "notify"]):
                synergies.append({
                    "local_name": local["name"],
                    "local_type": local["type"],
                    "description": f"Target retrieves data → {local['name']} can send/notify"
                })
            elif any(kw in target_desc for kw in ["search", "find"]) and \
                 any(kw in local_desc for kw in ["analyze", "summarize", "process"]):
                synergies.append({
                    "local_name": local["name"],
                    "local_type": local["type"],
                    "description": f"Target searches → {local['name']} can analyze results"
                })
        return synergies

    def suggest_install(self, overlap_results: List[Dict], synergy_results: List[Dict]) -> Tuple[str, str]:
        """Return (recommendation, reasoning)."""
        high_overlap = any(r["level"] == "high" for r in overlap_results)
        partial_overlap = any(r["level"] == "partial" for r in overlap_results)
        if high_overlap:
            return "❌ Do not install", "High overlap with existing capability"
        if synergy_results:
            return "✅ Recommended", "Can work synergistically with existing tools"
        if partial_overlap:
            return "⚠️ Consider carefully", "Partial overlap; may extend or conflict"
        return "✅ Recommended", "No conflict and no overlap"

File: src/matcher.ts
import type { WebpageInfo } from "./webfetch.js";
import type { LocalCapability } from "./scanner.js";

export interface OverlapItem {
  name: string;
  type: string;
  agent: string;
  similarity: number;
  level: string;
  reason: string;
}

export interface SynergyItem {
  name: string;
  type: string;
  description: string;
}

export interface AnalysisResult {
  overlap: OverlapItem[];
  synergy: SynergyItem[];
  recommendation: string;
  reason: string;
}

// Simple TF-IDF like implementation
function tokenize(text: string): string[] {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9\s]/g, " ")
    .split(/\s+/)
    .filter((word) => word.length > 2);
}

function termFrequency(term: string, tokens: string[]): number {
  const count = tokens.filter((t) => t === term).length;
  return count / tokens.length;
}

function inverseDocumentFrequency(term: string, documents: string[][]): number {
  const docsWithTerm = documents.filter((doc) => doc.includes(term)).length;
  if (docsWithTerm === 0) return 0;
  return Math.log(documents.length / docsWithTerm);
}

function computeTfidfSimilarity(text1: string, text2: string): number {
  const tokens1 = tokenize(text1);
  const tokens2 = tokenize(text2);
  
  if (tokens1.length === 0 || tokens2.length === 0) {
    return 0;
  }

  const allTokens = [tokens1, tokens2];
  const uniqueTerms = [...new Set([...tokens1, ...tokens2])];

  // Compute TF-IDF vectors
  const tfidf1: number[] = [];
  const tfidf2: number[] = [];

  for (const term of uniqueTerms) {
    const tf1 = termFrequency(term, tokens1);
    const tf2 = termFrequency(term, tokens2);
    const idf = inverseDocumentFrequency(term, allTokens);
    
    tfidf1.push(tf1 * idf);
    tfidf2.push(tf2 * idf);
  }

  // Cosine similarity
  const dotProduct = tfidf1.reduce((sum, val, i) => sum + val * tfidf2[i], 0);
  const magnitude1 = Math.sqrt(tfidf1.reduce((sum, val) => sum + val * val, 0));
  const magnitude2 = Math.sqrt(tfidf2.reduce((sum, val) => sum + val * val, 0));

  if (magnitude1 === 0 || magnitude2 === 0) {
    return 0;
  }

  return dotProduct / (magnitude1 * magnitude2);
}

function explainOverlap(target: WebpageInfo, local: LocalCapability, sim: number): string {
  if (sim > 0.85) {
    return "Likely duplicate functionality";
  } else if (sim > 0.6) {
    return "Partial functional overlap";
  }
  return "No significant overlap";
}

export function analyzeOverlap(
  target: WebpageInfo,
  localCapabilities: LocalCapability[]
): AnalysisResult {
  const targetText = `${target.name} ${target.description || ""}`;
  
  const overlap: OverlapItem[] = [];
  
  for (const local of localCapabilities) {
    const localText = `${local.name} ${local.description || ""}`;
    const sim = computeTfidfSimilarity(targetText, localText);
    
    let level: string;
    if (sim > 0.85) {
      level = "high";
    } else if (sim > 0.6) {
      level = "partial";
    } else {
      level = "none";
    }

    if (sim > 0.3) {
      overlap.push({
        name: local.name,
        type: local.type,
        agent: local.agent,
        similarity: Math.round(sim * 1000) / 1000,
        level,
        reason: explainOverlap(target, local, sim),
      });
    }
  }

  // Sort by similarity
  overlap.sort((a, b) => b.similarity - a.similarity);

  // Synergy detection (simple keyword-based)
  const synergy: SynergyItem[] = [];
  const targetDesc = (target.description || "").toLowerCase();
  
  for (const local of localCapabilities) {
    const localDesc = (local.description || "").toLowerCase();
    
    // Pipeline: fetch → send/notify
    if (
      (targetDesc.includes("fetch") || targetDesc.includes("get") || targetDesc.includes("read") ||
       targetDesc.includes("list") || targetDesc.includes("query")) &&
      (localDesc.includes("send") || localDesc.includes("post") || localDesc.includes("write") ||
       localDesc.includes("create") || localDesc.includes("notify"))
    ) {
      synergy.push({
        name: local.name,
        type: local.type,
        description: `Target retrieves data → ${local.name} can send/notify`,
      });
    }
    // Search → analyze
    else if (
      (targetDesc.includes("search") || targetDesc.includes("find")) &&
      (localDesc.includes("analyze") || localDesc.includes("summarize") || localDesc.includes("process"))
    ) {
      synergy.push({
        name: local.name,
        type: local.type,
        description: `Target searches → ${local.name} can analyze results`,
      });
    }
  }

  // Generate recommendation
  const highOverlap = overlap.some((o) => o.level === "high");
  const partialOverlap = overlap.some((o) => o.level === "partial");

  let recommendation: string;
  let reason: string;

  if (highOverlap) {
    recommendation = "Do not install";
    reason = "High overlap with existing capability";
  } else if (synergy.length > 0) {
    recommendation = "Recommended";
    reason = "Can work synergistically with existing tools";
  } else if (partialOverlap) {
    recommendation = "Consider carefully";
    reason = "Partial overlap; may extend or conflict";
  } else {
    recommendation = "Recommended";
    reason = "No conflict and no overlap";
  }

  return {
    overlap,
    synergy,
    recommendation,
    reason,
  };
}

export function suggestInstall(
  overlap: OverlapItem[],
  synergy: SynergyItem[]
): [string, string] {
  const result = analyzeOverlap(
    { name: "", type: "", url: "", tools: [], install_methods: [], tags: [] },
    []
  );
  return [result.recommendation, result.reason];
}